src/smt.rs

```diff
@@ -338,32 +346,34 @@ fn calc_next_info(info: &mut QueryHashesInfo, next_info: &mut NextQueryHashesInf
 }
 
 fn calculate_query_hashes(mut info: QueryHashesInfo) {
-    if info.extra.height.is_equal_to(0) {
-        return;
-    }
-    let mut next_info = NextQueryHashesInfo::new(info.extra.target_id);
-    let mut i = 0;
-    while i < info.layer_nodes.len() {
-        if info.layer_structure[i] != info.extra.height.into() {
-            next_info.push(info.layer_nodes[i].clone(), info.layer_structure[i]);
-            i += 1;
-            continue;
+    let mut is_extra_height_zero = info.extra.height.is_equal_to(0);
+
+    while !is_extra_height_zero {
+        let mut next_info = NextQueryHashesInfo::new(info.extra.target_id);
+        let mut i = 0;
+        while i < info.layer_nodes.len() {
+            if info.layer_structure[i] as u8 != info.extra.height.into() {
```

Comment:
`Height` has `From<u8>` so there is no needs to cast `layer_structure` to `u8`

Reply:
Done.